extensions/amp-bind/0.1/amp-state.js

```diff
@@ -207,15 +210,34 @@ export class AmpState extends AMP.BaseElement {
   /**
    * @param {boolean} isInit
    * @param {boolean=} opt_refresh
-   * @return {!Promise<undefined>}
+   * @return {!Promise}
    * @private
    */
   fetchAndUpdate_(isInit, opt_refresh) {
     // Don't fetch in prerender mode.
-    return this.getAmpDoc()
+    const fetchAndUpdatePromise = this.getAmpDoc()
       .whenFirstVisible()
       .then(() => this.prepareAndSendFetch_(isInit, opt_refresh))
       .then(json => this.updateState_(json, isInit));
+
+    this.fetchAndUpdatePromise_ = fetchAndUpdatePromise;
+
+    // Cleanup when complete
+    fetchAndUpdatePromise.then(() => {
+      if (this.fetchAndUpdatePromise_ === fetchAndUpdatePromise) {
+        this.fetchAndUpdatePromise_ = null;
+      }
+    });
+
+    return fetchAndUpdatePromise;
+  }
+
+  /**
+   * Returns a promise that resolves after the fetch and update completes.
+   * @return {Promise}
+   */
+  getFetchAndUpdatePromise() {
+    return this.fetchAndUpdatePromise_;
```

Comment:
This function returning `null` has multiple meanings which may be confusing -- amp-state is done fetching, or amp-state has nothing to fetch, or amp-state is not yet built. How could we make this less error prone?

Reply:
I dont think "amp state is done fetching" and "amp state has nothing to fetch" are cases worth distinguishing between.  built vs. not-built is. I'll take a look

Reply:
I've made it so that this function now always returns a Promise:
- A resolved promise for the local json case (no src attr)
- A Promise that resolves when the fetch does. It works in the not-yet-built case as well